### PR TITLE
Updated ReadWriteSpinlock to be atomically upgrade-able from Reader to

### PR DESCRIPTION
**Pull request by accelerated**
### Description of changes

Updated `ReadWriteSpinlock` to be atomically upgrade-able from Reader to Writer.

Signed-off-by: Alexander Damian <adamian@bloomberg.net>

### CODE REVIEW

File: quantum/util/impl/quantum_spinlock_util_impl.h [hunk 1]
```diff
@@ -216,4 +421,24 @@ size_t& SpinLockUtil::numSpins()
     return _numSpins;
 }
 
+inline constexpr
+uint32_t SpinLockUtil::set(int16_t h, int16_t l) {
+    return ((uint32_t)h<<16) | (l&mask);
+}
+
+inline constexpr
+int16_t SpinLockUtil::getHigh(uint32_t n) {
```
Comment (arosenzweig3) [inline quantum/util/impl/quantum_spinlock_util_impl.h:430]: While this is accurate, I'm tempted to say we should rename getHigh & getLow to something more indicative of the values stored in the high & low halves of the int; it's very easy when reading this code to forget which value is stored in which half.

File: tests/quantum_locks_tests.cpp [hunk 2]
```diff
@@ -351,6 +354,100 @@ TEST(ReadWriteSpinLock, Guards)
...
+    lock.lockRead();
+    lock.lockRead();
+    lock.lockRead();
+    EXPECT_TRUE(lock.isLocked());
+    EXPECT_TRUE(lock.isReadLocked());
+    EXPECT_FALSE(lock.isWriteLocked());
+    EXPECT_EQ(3, lock.numReaders());
+    EXPECT_EQ(0, lock.numPendingWriters());
+    std::thread t([&lock]() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (arosenzweig3) [inline tests/quantum_locks_tests.cpp:369]: This part of the test looks very timing-specific and thus fragile. Let's find a more robust method of coordination; perhaps check out some of the tests in my PR in the private quantum?
My concern is that this basically assumes that if the thread sleeps 100ms, the upgradeToWrite() call will have started. That should be true, but if for some reason the system is slow, the test fails.
Perhaps we could just have the thread do
while (lock.numPendingWriters() == 0) {} instead of the sleep

File: tests/quantum_locks_tests.cpp [hunk 3]
```diff
@@ -351,6 +354,100 @@ TEST(ReadWriteSpinLock, Guards)
...
+    EXPECT_FALSE(lock.tryUpgradeToWrite()); //2 readers is impossible
+    EXPECT_FALSE(lock.isWriteLocked());
+    lock.unlockRead(); //only a single reader left
+    EXPECT_EQ(1, lock.numReaders());
+    
+    //start a bunch of parallel readers which will then upgrade to writers
+    std::vector<std::thread> threads;
+    auto timeToWake = std::chrono::system_clock::now() + std::chrono::milliseconds(10);
+    for (int i = 0; i < 10; ++i) {
+        threads.emplace_back([&lock, &timeToWake]() {
```
Comment (arosenzweig3) [inline tests/quantum_locks_tests.cpp:430]: Similarly let's look at a non-time-based coordination method.

File: quantum/impl/quantum_read_write_spinlock_impl.h [hunk 4]
```diff
@@ -156,6 +180,23 @@ bool ReadWriteSpinLock::ReadGuard::tryLock()
     return _ownsLock;
 }
 
+inline
+void ReadWriteSpinLock::ReadGuard::upgradeToWrite()
+{
+    if (!_ownsLock) {
+        _spinlock.upgradeToWrite();
```
[the remark this answers is not in the document]
Comment (accelerated) [reply quantum/impl/quantum_read_write_spinlock_impl.h:187]: Yes good point...will make the change

File: quantum/impl/quantum_read_write_spinlock_impl.h [hunk 5]
```diff
@@ -156,6 +180,23 @@ bool ReadWriteSpinLock::ReadGuard::tryLock()
...
+    }
+}
+
+inline
+bool ReadWriteSpinLock::ReadGuard::tryUpgradeToWrite()
+{
+    if (_ownsLock) {
+        return _spinlock.tryUpgradeToWrite();
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (accelerated) [reply quantum/impl/quantum_read_write_spinlock_impl.h:197]: Ok let's leave it as a no-op instead of an assert. The return code gives you an indication as well.

File: tests/quantum_locks_tests.cpp [hunk 6]
```diff
@@ -351,6 +354,150 @@ TEST(ReadWriteSpinLock, Guards)
...
+    lock.unlockWrite();
+    for (auto&& t : threads) {
+        t.join();
+    }
+    EXPECT_EQ(0, lock.numReaders());
+    EXPECT_EQ(0, lock.numPendingWriters());
+    EXPECT_FALSE(lock.isLocked());
+    EXPECT_EQ(-1, values.front());
+}
+
```
Comment (arosenzweig3) [inline tests/quantum_locks_tests.cpp:500]: Please also add tests of the guards, especially the ReadGuard upgrading to a write lock and correctly unlocking.

File: quantum/util/impl/quantum_spinlock_util_impl.h (hunk 1 above)
Comment (accelerated) [reply quantum/util/impl/quantum_spinlock_util_impl.h:429]: Ok renamed to upgrades() and owners() respectively.

File: quantum/impl/quantum_mutex_impl.h [hunk 7]
```diff
@@ -115,9 +119,8 @@ inline
 Mutex::Guard::Guard(Mutex& mutex,
                     Mutex::AdoptLock) :
     _mutex(&mutex),
-    _ownsLock(true)
+    _ownsLock(mutex.isLocked())
```
Comment (arosenzweig3) [inline quantum/impl/quantum_mutex_impl.h:122]: Note for the future: we should probably have an assert here that verifies that the guard owner is the mutex owner; that requires the mutex itself to offer an ownsLock(TaskId = local::taskId()) API. Let's do that next time.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 8]
```diff
@@ -121,20 +144,29 @@ int ReadWriteMutex::numReaders() const
...
+}
+
 //==============================================================================================
 //                                class ReadGuard
 //==============================================================================================
 inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock) :
     ReadWriteMutex::ReadGuard::ReadGuard(nullptr, lock)
 {
+    //Application must use the other lock() overload if we are running inside a coroutine
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:160]: Comment should reference constructor, not lock()

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 9]
```diff
@@ -143,35 +175,40 @@ inline
...
+    _isUpgraded(false)
 {
 }
 
 inline
 ReadWriteMutex::ReadGuard::ReadGuard(ReadWriteMutex& lock,
                                      ReadWriteMutex::AdoptLock) :
     _mutex(&lock),
-    _ownsLock(true)
+    _ownsLock(lock.isReadLocked()),
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:187]: This looks wrong; if we adopt an upgraded lock, _ownsLock will be false but _isUpgraded will be true. Then we won't unlock in the destructor.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 10]
```diff
@@ -143,35 +175,40 @@ inline
...
+    _isUpgraded(lock.isWriteLocked())
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:188]: Similarly to Mutex::Guard, we should look at adding APIs to ReadWriteMutex to identify if the caller owns the lock, and assert based on those APIs for the guard AdoptLock constructors. This has to wait for the fully owner-aware ReadWriteMutex.

File: quantum/impl/quantum_read_write_spinlock_impl.h [hunk 11]
```diff
@@ -27,147 +64,178 @@ namespace quantum {
...
+    _isUpgraded(false)
+{
+}
+
+inline
+ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock,
+                                        ReadWriteSpinLock::AdoptLock) :
     _spinlock(lock),
-    _ownsLock(_spinlock.tryLockRead())
+    _ownsLock(lock.isReadLocked()),
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_spinlock_impl.h:164]: Similar to the ReadWriteMutex::ReadGuard(lock, AdoptLock), the logic on _ownsLock here is wrong.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 12]
```diff
@@ -180,24 +217,54 @@ inline
...
     
-    if (_ownsLock) return true;
-    assert(_mutex);
+    assert(_mutex && !ownsLock());
     _ownsLock = _mutex->tryLockRead();
     return _ownsLock;
 }
 
+inline
+void ReadWriteMutex::ReadGuard::upgradeToWrite()
```
Comment (demin80) [inline quantum/impl/quantum_read_write_mutex_impl.h:226]: With the upgradeToWrite feature of ReadGuard, we don't seem to need WriteGuard because it can be achieved via
ReadGuard g(...);
g.upgradeToWrite();
Is that so? 

Alternatively, instead of adding ReadGuard::upgradeToWrite(), maybe add a ctor to WriteGuard taking ReadGuard, that will steal the mutex from the ReadGuard and upgrade it to write?

File: quantum/impl/quantum_read_write_spinlock_impl.h [hunk 13]
```diff
@@ -18,6 +18,43 @@
 //##############################################################################################
 //#################################### IMPLEMENTATIONS #########################################
 //##############################################################################################
+/*
+The atomic 32-bit counter is divided into two halves, the high 16 bits representing the number of pending
+writer upgrades and the low 16 bit representing the current number of reader owners or -1
```
Comment (demin80) [inline quantum/impl/quantum_read_write_spinlock_impl.h:23]: "representing the current number of reader owners or -1" why is it -1 but not 0? What does 0 mean currently?

File: quantum/impl/quantum_read_write_mutex_impl.h (hunk 12 above)
Comment (arosenzweig3) [reply quantum/impl/quantum_read_write_mutex_impl.h:226]: I do actually like the semantics of Denis' proposal here; WriteGuard(ReadGuard&, TryToLock{}) might be a bit awkward to implement but the semantics feel a lot better than a ReadGuard that owns a write lock.
We definitely want to have a way to go straight to write lock via guards for users who know they're going to need the write lock and don't want to pay the price of read lock + upgrade.

File: quantum/impl/quantum_read_write_spinlock_impl.h (hunk 13 above)
Comment (accelerated) [reply quantum/impl/quantum_read_write_spinlock_impl.h:23]: -1 means write locked. 0 means unlocked and >0 means read locked. I will update the comment better.

File: quantum/impl/quantum_read_write_mutex_impl.h (hunk 12 above)
Comment (accelerated) [reply quantum/impl/quantum_read_write_mutex_impl.h:226]: @demin80, yes however the ReadGuard vs WriteGuard indicates intent more clearly. If your code only needs to write (i.e. a producer) then you can only use a WriteGuard. With the WriteGuard taking a ReadGuard and upgrading I feel you have to create 2 guard objects instead of one.

Comment (accelerated) [reply quantum/impl/quantum_read_write_mutex_impl.h:226]: I personally prefer either the two distinct guards (with the reader having an upgrade) or a single guard which can upgrade as needed.

File: quantum/impl/quantum_read_write_spinlock_impl.h (hunk 13 above)
Comment (accelerated) [reply quantum/impl/quantum_read_write_spinlock_impl.h]: Ok I improved description.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 14]
```diff
@@ -121,177 +144,179 @@ int ReadWriteMutex::numReaders() const
...
-    _ownsLock = false;
-    _mutex = nullptr;
+    if (ownsLock()) {
+        unlock();
+    }
 }
 
 inline
-bool ReadWriteMutex::ReadGuard::ownsLock() const
+void ReadWriteMutex::Guard::lock()
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:234]: I think we should have a lockRead() and a lockWrite() here.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 15]
```diff
@@ -121,177 +144,179 @@ int ReadWriteMutex::numReaders() const
...
+    _mutex->lockRead(sync);
+    _ownsLock = true;
 }
 
 inline
-ReadWriteMutex::WriteGuard::WriteGuard(ICoroSync::Ptr sync,
-                                       ReadWriteMutex& lock) :
-    _mutex(&lock),
-    _ownsLock(true)
+bool ReadWriteMutex::Guard::tryLock()
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:250]: likewise tryLockRead() & tryLockWrite() instead of lock()

File: quantum/impl/quantum_read_write_spinlock_impl.h [hunk 16]
```diff
@@ -18,213 +18,257 @@
...
+    if (ownsLock()) {
+        unlock();
     }
 }
 
 inline
-ReadWriteSpinLock::WriteGuard::WriteGuard(ReadWriteSpinLock& lock) :
-    _spinlock(lock),
-    _ownsLock(true)
+void ReadWriteSpinLock::Guard::lock()
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_spinlock_impl.h:210]: Same comment as on ReadWriteMutex::Guard; we need lockRead() & lockWrite() rather than lock().

File: quantum/quantum_read_write_mutex.h [hunk 17]
```diff
@@ -93,32 +102,48 @@ class ReadWriteMutex
...
+    
+    /// @brief Returns the number of pending upgraded writers.
+    /// @return The number of writers.
+    int numPendingWriters() const;
 
-    class ReadGuard
+    class Guard
     {
     public:
         /// @brief Construct this object and lock the passed-in mutex as a reader.
```
Comment (arosenzweig3) [inline quantum/quantum_read_write_mutex.h:113]: Delete 'as a reader' here.

File: quantum/quantum_read_write_mutex.h [hunk 18]
```diff
@@ -93,32 +102,48 @@ class ReadWriteMutex
...
+        /// @brief Construct this object and assumes the current state of the lock w/o modifying it.
         /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
-        /// @note Attempts to lock the mutex. Does not block.
-        ReadGuard(ReadWriteMutex& lock,
-                  ReadWriteMutex::AdoptLock);
+        /// @note Does not block.
+        Guard(ReadWriteMutex& lock,
+              LockTraits::AdoptLock);
         
         /// @brief Destroy this object and unlock the underlying mutex if this object owns it.
```
Comment (arosenzweig3) [inline quantum/quantum_read_write_mutex.h:145]: Rephrase this as "unlock the underlying mutex if this object still owns a lock state on it" or something similar. That is, the object doesn't really give up ownership of the Mutex object (unless we implement release() again). It always owns the Mutex, what it may or may not own is the locked state.

File: quantum/quantum_read_write_mutex.h [hunk 19]
```diff
@@ -129,59 +154,20 @@ class ReadWriteMutex
...
+        /// @brief Try to upgrade this reader to a writer.
+        /// @return True is succeeded.
         /// @note Does not block.
-        bool tryLock();
+        /// @note The lock must already be owned prior to invoking this function.
+        bool tryUpgradeToWrite();
 
-        /// @brief Releases the write lock on the underlying mutex.
+        /// @brief Releases the read lock on the underlying mutex.
         /// @note Also releases ownership of the underlying mutex.
```
Comment (arosenzweig3) [inline quantum/quantum_read_write_mutex.h:171]: This comment isn't quite right (I expect it's replicated on others); see comments above on destructor.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 20]
```diff
@@ -121,177 +144,203 @@ int ReadWriteMutex::numReaders() const
...
 
-//==============================================================================================
-//                                class WriteGuard
-//==============================================================================================
 inline
-ReadWriteMutex::WriteGuard::WriteGuard(ReadWriteMutex& lock) :
-    ReadWriteMutex::WriteGuard::WriteGuard(nullptr, lock)
+void ReadWriteMutex::Guard::lockWrite(ICoroSync::Ptr sync)
 {
+    assert(_mutex && !ownsLock());
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:261]: should assert !ownsWriteLock() as well.

File: quantum/impl/quantum_read_write_mutex_impl.h [hunk 21]
```diff
@@ -121,177 +144,203 @@ int ReadWriteMutex::numReaders() const
...
 
 inline
-ReadWriteMutex::WriteGuard::WriteGuard(ReadWriteMutex& lock,
-                                       ReadWriteMutex::TryToLock) :
-    _mutex(&lock),
-    _ownsLock(_mutex->tryLockWrite())
+bool ReadWriteMutex::Guard::tryLockWrite()
 {
+    
+    assert(_mutex && !ownsLock());
```
Comment (arosenzweig3) [inline quantum/impl/quantum_read_write_mutex_impl.h:279]: assert (!ownsWriteLock()) as well

File: quantum/impl/quantum_read_write_mutex_impl.h (hunk 20 above)
Comment (accelerated) [reply quantum/impl/quantum_read_write_mutex_impl.h:261]: ownsLock includes ownsWriteLock.

Comment (arosenzweig3) [reply quantum/impl/quantum_read_write_mutex_impl.h:261]: my mistake, this (and others) looks good.
